File: services/orders/app/dao/order.py
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import (
    String,
    DateTime,
    Numeric,
    Enum as SAEnum,
    func,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base
from app.dao.order_status import OrderStatus


class Order(Base):
    __tablename__ = "orders"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    customer_name: Mapped[str] = mapped_column(String(255), nullable=False)
    status: Mapped[OrderStatus] = mapped_column(
        SAEnum(OrderStatus, name="order_status", create_constraint=True),
        default=OrderStatus.PENDING,
        index=True,
    )
    total: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0)
    user_id: Mapped[str] = mapped_column(
        String(255), nullable=False, index=True, comment="Email do usuário autenticado"
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    items: Mapped[list["OrderItem"]] = relationship(
        back_populates="order", cascade="all, delete-orphan", lazy="selectin"
    )

File: frontend/mfe-orders/src/types/order.ts
export type OrderStatus = 'pendente' | 'confirmado' | 'enviado' | 'entregue' | 'cancelado'

export interface OrderItem {
  id?: string
  product_name: string
  quantity: number
  unit_price: number
}

export interface Order {
  id: string
  customer_name: string
  status: OrderStatus
  total?: number
  user_id?: string
  items: OrderItem[]
  created_at?: string
  updated_at?: string
}

export interface UpdateOrderStatusPayload {
  status: OrderStatus
}

export interface CreateOrderPayload {
  customer_name: string
  items: Array<{
    product_name: string
    quantity: number
    unit_price: number
  }>
}
